### PR TITLE
Sort by mutation time and check sort in integrity

### LINKED ISSUE #651: Add requirement that mutations are sorted by time
> A though: once this is implemented, we should have `sort_tables` sort mutations by time, which satisfies the current requirement that parents come before children, but would make the ordering produced by sort_tables less ambiguous (not totally, if in discrete time, though). Perhaps we should make "sorted by time" a requirement of the mutation table, also.

_Originally posted by @petrelharp in https://github.com/tskit-dev/tskit/issues/513#issuecomment-634211466_

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #651 

Needs python test work now.

### CODE REVIEW

File: c/tskit/tables.c [hunk 1]
```diff
@@ -6742,6 +6747,23 @@ tsk_table_collection_check_mutation_integrity(
                     parent_mut = mutations.parent[parent_mut];
                 }
             }
+            
+            /* Check time ordering, we do this after the time checks above, so that
+               more specific errors trigger first */
+            if (!tsk_is_unknown_time(last_known_time) &&
```
Comment (jeromekelleher) [inline c/tskit/tables.c:6753]: It's a bit confusing to have to check ``last_known_time`` is known - could we avoid this by setting ``last_known_time`` to -1 initially or something?

Comment (jeromekelleher) [reply c/tskit/tables.c:6753]: We've also already checked for ``!unknown_time`` just above - why not keep this in the same block?

Comment (benjeffery) [reply c/tskit/tables.c:6753]: Negative times are valid, so using unknown seemed the best way to say "no known time has been seen for this site".

Comment (jeromekelleher) [reply c/tskit/tables.c:6753]: Aha, so they are. What about -inf then? This is always less than everything, isn't it, and we check for ``isfinite(time)`` before this.

Comment (benjeffery) [reply c/tskit/tables.c]: I've used positive infinity (parents before children right?) in https://github.com/tskit-dev/tskit/pull/726/commits/3057d9897f39cd252170a256005f2a41b03cc1e2

File: c/tskit/tables.c [hunk 2]
```diff
@@ -6742,6 +6747,23 @@ tsk_table_collection_check_mutation_integrity(
...
+                !unknown_time &&
+                j > 0 &&
+                mutations.site[j - 1] == mutations.site[j] &&
+                mutation_time > last_known_time
+            ){
+                    ret = TSK_ERR_UNSORTED_MUTATIONS;
+                    goto out;
+            }
+
+            if (!unknown_time || (j>0 && mutations.site[j - 1] != mutations.site[j])) {
```
[the remark this answers is not in the document]
Comment (benjeffery) [reply c/tskit/tables.c]: I've modified the tests to cover this sloppy thinking! Fixed in https://github.com/tskit-dev/tskit/pull/726/commits/3057d9897f39cd252170a256005f2a41b03cc1e2

File: c/tskit/tables.c [hunk 3]
```diff
@@ -6719,6 +6724,10 @@ tsk_table_collection_check_mutation_integrity(
                 }
             }
 
+            if (j > 0 && mutations.site[j - 1] != mutations.site[j]) {
```
Comment (jeromekelleher) [inline c/tskit/tables.c:6727]: I'm nitpicking here, but we could avoid the j > 0 comparison and be a bit more cache friendly if we put this check into the (j > 0) block around 6706. It probably doesn't matter, but we do run this every time we create a tree sequence object, so it makes sense to keep it as tight as we can.;

Comment (benjeffery) [reply c/tskit/tables.c]: Nah, not a nickpit! Moved in https://github.com/tskit-dev/tskit/pull/726/commits/79dd9ecd10227631a7dec75c722fb109bb98cc9b

**Comment by benjeffery** [comment]
@jeromekelleher RFR! Happy to hear suggestions for further tests or docs I missed.

File: c/tskit/core.c [hunk 4]
```diff
@@ -286,7 +286,9 @@ tsk_strerror_internal(int err)
             ret = "Inconsistent mutations: state already equal to derived state";
             break;
         case TSK_ERR_UNSORTED_MUTATIONS:
-            ret = "Mutations must be provided in non-decreasing site order";
+            ret = "Mutations must be provided in non-decreasing site order and "
+                  "decreasing "
```
Comment (jeromekelleher) [inline c/tskit/core.c:290]: Join lines here

**Comment by benjeffery** [comment]
Actually, hold fire - seems windows wants to put unknown first and `glibc` wants to put it last.

File: docs/data-model.rst [hunk 5]
```diff
@@ -653,7 +653,8 @@ Furthermore,
 For simplicity and algorithmic efficiency, mutations must also:
 
 - be sorted by site ID;
-- when there are multiple mutations per site, parent mutations must occur
+- when there are multiple mutations per site, mutations should be ordered by
+  decreasing time, if known, and parent mutations must occur
```
Comment (jeromekelleher) [inline docs/data-model.rst:657]: s/decreasing/nonincreasing

Adjacent times can be equal, right? Also in the changelong.

File: python/tests/test_highlevel.py [hunk 6]
```diff
@@ -754,6 +754,7 @@ def test_compute_mutation_time(self):
                 np.allclose(python_time, tables.mutations.time, rtol=1e-15, atol=1e-15)
             )
             # Check we have valid times
+            tables.sort()
```
Comment (jeromekelleher) [inline python/tests/test_highlevel.py:757]: Why do we need to sort here? This would be a bug in compute_mutation_times, wouldn't it?

**Comment by benjeffery** [comment]
So for the current `cmp_mutation` in this PR if you have three mutations with times 4, UNK, 5 (and hence ids 0,1,2) then you get:
5<4, 4<U, U<5 which is inconsistent. With glibc this results in 4,5,U and windows does U,4,5. We could fix this by saying unknown times are either + or - infinity. BUT that might mean we broke any ordering that was present due to the "parents first" rule. I'm not sure how to fix this as check for a parent chain in `cmp_mutation` sounds expensive.

File: python/tests/test_highlevel.py (hunk 6 above)
Comment (benjeffery) [reply python/tests/test_highlevel.py:757]: This is failing on the example tree from `insert_branch_mutations` looking into it.

Comment (benjeffery) [reply python/tests/test_highlevel.py:757]: Massive confusion till I realised this test uses a non-fixed random seed! :facepalm:

Comment (jeromekelleher) [reply python/tests/test_highlevel.py:757]: Dammit, I hate it when that happens! :smile:

**Comment by jeromekelleher** [comment]
Is this because of stable and unstable sorting algs in the libc implementations @benjeffery?

File: python/tests/test_highlevel.py (hunk 6 above)
Comment (benjeffery) [reply python/tests/test_highlevel.py:757]: Here's the badly sorted data that's come out of `compute_mutation_times`. Just a single site. I'm trying to get my head round what's going on. https://gist.github.com/benjeffery/fe7727ae93bbe79fa5b3509428097465

Comment (benjeffery) [reply python/tests/test_highlevel.py:757]: I don't think this is a bug in `compute_mutation_times` The order going in is just "sites and parents before children" but once times are applied (using the node times) the old order is no longer valid.

File: c/tskit/core.c (hunk 4 above)
Comment (benjeffery) [reply c/tskit/core.c]: https://github.com/tskit-dev/tskit/pull/726/commits/4427cd24a7551634867d4d2656f3938389a3fa0c

File: docs/data-model.rst (hunk 5 above)
Comment (benjeffery) [reply docs/data-model.rst:657]: https://github.com/tskit-dev/tskit/pull/726/commits/4427cd24a7551634867d4d2656f3938389a3fa0c

File: c/tskit/tables.c [hunk 7]
```diff
@@ -4343,9 +4343,12 @@ cmp_mutation(const void *a, const void *b)
     const tsk_mutation_t *ib = (const tsk_mutation_t *) b;
     /* Compare mutations by site */
     int ret = (ia->site > ib->site) - (ia->site < ib->site);
+    /* Within a particular site sort by time if known, then ID. This ensures that
+     * relative ordering within a site is maintained */
+    if (ret == 0 && !tsk_is_unknown_time(ia->time) && !tsk_is_unknown_time(ib->time)) {
```
Comment (jeromekelleher) [inline c/tskit/tables.c:4348]: I can't see how this wouldn't completely determine the sort order - if either time is unknown we should fall back on comparing by ID.

Comment (benjeffery) [reply c/tskit/tables.c:4348]: The problem here is that our `cmp` gives inconsistent orderings (5<4, 4<U, U<5) in my example above as one is by time  and the other two by id. The cycle has to be broken somehow.

Comment (benjeffery) [reply c/tskit/tables.c:4348]: @petrelharp I think you might be helpful here as I'm not sure how to fix this one! This comment hopefully explains the situation: https://github.com/tskit-dev/tskit/pull/726#issuecomment-662429215

**Comment by jeromekelleher** [comment]
Ugh, this is nasty. The only simple way out of this I can see is to disallow mixed known and unknow times at a site. That way we should never have to compare unknown times with anything. It's the edgiest of edge-cases, in any case (who is actually going to do this?).

@petrelharp, any thoughts here?

It's great this has come up now - testing works!

**Comment by jeromekelleher** [comment]
@petrelharp, we could use your input here. Do you object to disallowing mixed known and unknown times at a site? Otherwise, we get into undefined behaviour because or sort keys are ill defined, as Ben has shown here.

File: python/tests/test_highlevel.py (hunk 6 above)
Comment (jeromekelleher) [reply python/tests/test_highlevel.py:757]: So I think it is a bug in compute_mutation_times if it doesn't return a well-formed mutation table. If we have to, we could do a site-wise sort, once all the times have been generated? (Or if this is too awkward, we could do a full TableCollection.sort(), but this is a very heavy weapon.)

Comment (benjeffery) [reply python/tests/test_highlevel.py]: In https://github.com/tskit-dev/tskit/pull/726/commits/30a3134d8c6ff5258fec539d5f260dd99361f503 I've used `tsk_table_collection_sort` but only if it is needed. If this turns out to be a bottleneck we can put more effort in, but I'm not sure it is worth the complexity without a pressing use case.

**Comment by petrelharp** [comment]
I agree: disallow mixed missing/nonmissing times. I don't see a good way around it and don't know if a reason we'd need to have mixed times. Ps Sorry for the delay, I am on semi vacation.

**Comment by benjeffery** [comment]
No worries, thanks for having a look. We can allow mixed known/unknown as long as `sort` isn't called. Or is it might be cleaner to have it be a table-level requirement.

**Comment by jeromekelleher** [comment]
> We can allow mixed known/unknown as long as sort isn't called. Or is it might be cleaner to have it be a table-level requirement.

Let's enforce it as a table-level requirement. It's just hassle for us otherwise.

File: python/tests/test_highlevel.py (hunk 6 above)
Comment (benjeffery) [reply python/tests/test_highlevel.py]: Need to change this to use a bookmark pointing to the end of the edge table to skip edge sorting.

Comment (benjeffery) [reply python/tests/test_highlevel.py]: Skip edge sorting in https://github.com/tskit-dev/tskit/pull/726/commits/36ce487bba0a3b9aa45be0f013840ab690f1a21b

**Comment by benjeffery** [comment]
Seems to be some issues with coverage reporting as C lines I know are covered from python are not showing as such.

**Comment by mufernando** [comment]
nice work on this @benjeffery, thanks!

**Pull request merged**